### PR TITLE
normalize telecom model_validator

### LINKED ISSUE #248: validate and normalize telecom data
## Summary
Normalize and validate all telecom data upon receipt, instead of when emitting the values for evaluation.

## Acceptance Criteria
- [x] Remove Telecom.phone and Telecom.email methods
- [x] Add a Telecom model validator to validate and normalize phone numbers
- [x] Add a Telecom model validator to validate email addresses (no need to normalize)
- [x] Update feature_iter to emit the appropriate email/phone values
- [x] Update site/reference.md to give guidance on how phone is normalized

## Details / Tasks
For phone number, please consider using [phonenumbers](https://pypi.org/project/phonenumbers/) or another lib to parse the incoming number and produce a [E164](https://en.wikipedia.org/wiki/E.164) string.

Validate email addresses with an email regex.

Both email addresses and phone numbers should use the `normalize_text` method in the feature_iter.

### PR DESCRIPTION
**Pull request by m-goggins**
## Description
This PR adds a model validator to the `Telecom` class that strips leading/trailing whitespace from emails and normalizes phone as an E164 string, i.e., including country codes but excluding extensions. When using `feature_iter`, only the national number (no country code) is used for comparison to avoid issues with Jaro-Winkler over indexing on similarities in country codes. For emails, we do not apply text normalization in `feature_iter` so that we can preserve the special characters for comparison. 

Storing:
- email: email with leading/trailing whitespace removed
- phone number: normalized to E164 where possible, otherwise store raw number

Comparing:
-email: same as storage
phone number: national number (no country code) and normalized_text() is applied to remove additional special characters. I primarily kept normalized_text in case we were not able to normalize to E164 during ingestion and we want to remove special characters.

## Related Issues
Fixes #248 

## Additional Notes

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: src/recordlinker/schemas/pii.py [hunk 1]
```diff
@@ -234,36 +234,31 @@ class Telecom(StrippedBaseModel):
...
-            name: name
-            for name, method in cls.__dict__.items()
-            if callable(method) and not name.startswith("_")
-        }
+    @pydantic.model_validator(mode="before")
+    def validate_and_normalize_telecom(cls, values):
+        """
+        Validate and normalize the telecom record.
+        """
+        # If telecom.system = "email", set telecom.value to lowercase
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:242]: Based on the comment, I think we're missing a call to `lower()`

File: src/recordlinker/schemas/pii.py [hunk 2]
```diff
@@ -441,27 +438,31 @@ def feature_iter(self, feature: Feature) -> typing.Iterator[str]:
                     yield str(race)
         elif attribute == FeatureAttribute.TELECOM:
             for telecom in self.telecom:
+                value = telecom.value.strip()
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:441]: I don't think we need to call `strip()`, all whitespace should have been removed by the model_validator

File: src/recordlinker/schemas/pii.py [hunk 3]
```diff
@@ -441,27 +438,31 @@ def feature_iter(self, feature: Feature) -> typing.Iterator[str]:
...
-                handlers = Telecom.get_system_handlers()
-
-                if telecom.system in handlers:
-                    value = getattr(telecom, handlers[telecom.system])()
-                    if value:
-                        yield value
-
+                elif telecom.system == "email":
+                    yield value
+                elif telecom.system == "phone":
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:447]: I think we just need 1 if/else block and no elif blocks here.  system == None or system == "email" is exactly the same as the else block.

Comment (m-goggins) [reply src/recordlinker/schemas/pii.py:447]: Ah you're right; forgot to clean it up after making other changes. Good catch!

File: src/recordlinker/schemas/pii.py (hunk 1 above)
Comment (m-goggins) [reply src/recordlinker/schemas/pii.py:242]: Oops, you're right

File: src/recordlinker/schemas/pii.py [hunk 4]
```diff
@@ -234,36 +234,31 @@ class Telecom(StrippedBaseModel):
...
+        Validate and normalize the telecom record.
+        """
+        # If telecom.system = "email", set telecom.value to lowercase
+        if values.get("system") == "email":
+            values["value"] = values["value"].strip().lower()
+        # If telecom.system = "phone", normalize the number
+        elif values.get("system") == "phone":
+            try:
+                # Attempt to parse with country code
+                if values["value"].startswith("+"):
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:249]: This is going to raise a server error if the user doesn't send the data in the correct format.  For example start up the local_server.sh and run this.

```sh
curl -X POST -H "Content-Type: application/json" -d '{"record": {"telecom": [{"system": "phone","val": " 555-123-4567"}]}}' http://localhost:8000/link
```

File: src/recordlinker/schemas/pii.py [hunk 5]
```diff
@@ -441,27 +438,25 @@ def feature_iter(self, feature: Feature) -> typing.Iterator[str]:
...
-
-                if telecom.system in handlers:
-                    value = getattr(telecom, handlers[telecom.system])()
-                    if value:
-                        yield value
-
+                value = telecom.value
+                if telecom.system == "phone":
+                    # Use national number for comparison
+                    phone = normalize_text(str(phonenumbers.parse(telecom.value).national_number))
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:444]:
```suggestion
                    phone = normalize_text(str(phonenumbers.parse(value).national_number))
```
Alternatively, the `value` local doesn't add much.  you could just change line 448 to `yield telecom.value`

**Review by ericbuckley — CHANGES_REQUESTED** [review]

File: src/recordlinker/schemas/pii.py (hunk 4 above)
Comment (m-goggins) [reply src/recordlinker/schemas/pii.py]: Ah, I switched the mode to "after", which should take care of this issue the model validator will now run after the whole model has been validated.

File: src/recordlinker/schemas/pii.py [hunk 6]
```diff
@@ -234,36 +234,32 @@ class Telecom(StrippedBaseModel):
...
+        # If telecom.system = "email", set telecom.value to lowercase
+        #
+        if self.system == "email":
+            self.value = self.value.strip().lower()
+        # If telecom.system = "phone", normalize the number
+        elif self.system == "phone":
+            try:
+                # Attempt to parse with country code
+                if self.value.startswith("+"):
+                    parsed_number = phonenumbers.parse(self.value)
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:251]: I don't think we're doing anything with the `parsed_number` variable.  Shouldn't we be assigning self.value to it?

File: src/recordlinker/schemas/pii.py [hunk 7]
```diff
@@ -441,27 +439,24 @@ def feature_iter(self, feature: Feature) -> typing.Iterator[str]:
...
-                handlers = Telecom.get_system_handlers()
-
-                if telecom.system in handlers:
-                    value = getattr(telecom, handlers[telecom.system])()
-                    if value:
-                        yield value
-
+                if telecom.system == "phone":
+                    # Use national number for comparison
+                    phone = normalize_text(str(phonenumbers.parse(telecom.value).national_number))
```
Comment (ericbuckley) [inline src/recordlinker/schemas/pii.py:444]: I understand why we are calling `normalize_text` here, but why are we reparsing the value with phonenumbers?  Shouldn't we only have to do that once during in the `validate_and_normalize_telecom` method?

Comment (ericbuckley) [reply src/recordlinker/schemas/pii.py:444]: Ah thanks for that explanation, that's helpful!

**Review by ericbuckley — APPROVED** [review]
🚀

File: tests/unit/schemas/test_pii.py [hunk 8]
```diff
@@ -583,3 +601,20 @@ def test_parse_state(self):
         assert address.state == "Armed Forces"
         address = pii.Address(state="Conneticut")
         assert address.state == "Conneticut"
+
+
+@pytest.mark.parametrize(
```
Comment (ericbuckley) [inline tests/unit/schemas/test_pii.py:606]: neat!

**Review by ericbuckley — APPROVED** [review]
